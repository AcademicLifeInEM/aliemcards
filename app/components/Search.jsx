import React from 'react';
import { post } from 'axios';
import { Link } from 'react-router';

class Search extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      query: '',
      cards: [],
    };
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(e) {
    const input = e.currentTarget.value;
    post('/graphql', {
      query: `query searchForCard($input: String){
        search(input: $input) {
          id
          title
        }
      }`,
      variables: { input },
    })
    .then(res => {
      if (res.status !== 200) throw res.status;
      console.log(res);
      const { search: cards } = res.data.data;
      this.setState({ cards, query: input });
    })
    .catch(err => console.error(`Error: API response status code = ${err}`));
  }

  render() {
    return (
      <div className="search">
        { this.props.hero &&
          <div className="search__splash-text">
            <p>A Point-of-Care reference library<br />
            by Michelle Lin, and the <strong>ALiEM Team</strong></p>
            <p>Formerly known as <em>Paucis Verbis Cards</em></p>
          </div>
        }
        <div className="search__input">
          <input type="text" onChange={this.handleChange} placeholder="Search" />
        </div>
        { this.state.cards.length > 0 &&
          <Results cards={this.state.cards} />
        }
      </div>
    );
  }
}

Search.propTypes = {
  hero: React.PropTypes.bool,
};
Search.defaultProps = {
  hero: false,
};

const Results = ({ cards }) => (
  <div className="container">
    <ul className="searchList">
      {cards.map((card) => (
        <li><Link to={`/cards/${card.id}`}>{card.title}</Link></li>
      ))}
    </ul>
  </div>
);

Results.propTypes = {
  cards: React.PropTypes.arrayOf(React.PropTypes.shape({
    id: React.PropTypes.string.isRequired,
    title: React.PropTypes.string.isRequired,
  })).isRequired,
};

export default Search;
