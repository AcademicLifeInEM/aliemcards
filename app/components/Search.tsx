import * as React from 'react';
import SearchResults from './SearchResults';

interface Props {
  splashText: boolean;
}

interface State {
  loading: boolean;
  query: string;
  uiQuery: string;
}

export default class Search extends React.PureComponent<Props, State> {
  static timer;

  constructor(props) {
    super(props);
    this.state = {
      loading: false,
      query: '',
      uiQuery: '',
    };
  }

  handleChange = (e: React.FormEvent<HTMLInputElement>) => {
    clearTimeout(Search.timer);
    const query = e.currentTarget.value;

    // Query is being performed. Update `uiQuery` immediately, but throttle
    // the update of `query` (and the associated api call) 500ms.
    if (query !== '') {
      Search.timer = setTimeout(() => {
        this.setState({ loading: false, query: this.state.uiQuery });
      }, 500);
      this.setState({ loading: true, uiQuery: query });
    }
    // Query is empty. Reset state.
    else {
      this.setState({ loading: false, uiQuery: query, query });
    }
  }

  handleClick = () => {
    this.setState({ query: '', uiQuery: '', loading: false });
  }

  render() {
    return (
      <div className="search">
        { this.props.splashText &&
          <div className="search__splash-text">
            <div>
              A Point-of-Care Reference Library
            </div>
            <div>
              by Michelle Lin, and the <strong>ALiEM Team</strong>
              <br />Formerly known as <em>Paucis Verbis Cards</em>
            </div>
          </div>
        }
        <div className="search__input" role="search">
          <input
            type="text"
            onChange={this.handleChange}
            placeholder="Search"
            aria-label="Search for cards"
            value={this.state.uiQuery}
          />
          { this.state.loading &&
            <img className="search__loader" src="/assets/images/loader.svg" alt="loader" />
          }
        </div>
        <SearchResults query={this.state.query} onClick={this.handleClick} />
      </div>
    );
  }
}
