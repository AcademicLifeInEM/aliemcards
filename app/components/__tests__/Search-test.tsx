jest.mock('react-router-dom');
jest.mock('../SearchResults');

import { shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Search from '../Search';

describe('<Search />', () => {
  it('should render correctly', () => {
    const component = renderer.create(<Search splashText={false} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shoulder render splashText when prop is set', () => {
    const component = renderer.create(<Search splashText={true} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have spinner only when searching', () => {
    jest.useFakeTimers();
    const component = shallow(<Search splashText={false} />);
    expect(component.find('.search__loader').exists()).toBeFalsy();
    component.find('input').simulate('change', { currentTarget: { value: 'hello' } });
    expect(component.find('.search__loader').exists()).toBeTruthy();
    jest.runAllTimers();
  });

  it('should not have spinner if query is empty', () => {
    const component = shallow(<Search splashText={false} />);
    expect(component.find('.search__loader').exists()).toBeFalsy();
    component.find('input').simulate('change', { currentTarget: { value: '' } });
    expect(component.find('.search__loader').exists()).toBeFalsy();
  });

  it('should clear query, uiQuery, and loader with handleClick', () => {
    const component = shallow(<Search splashText={false} />);
    component.find('input').simulate('change', { currentTarget: { value: 'hello' } });
    component.instance().handleClick();
    expect(component.state('loading')).toBe(false);
    expect(component.state('query')).toBe('');
    expect(component.state('uiQuery')).toBe('');
  });
});
