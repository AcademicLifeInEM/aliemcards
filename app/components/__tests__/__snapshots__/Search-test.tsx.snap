// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render correctly 1`] = `
<div
  className="search"
>
  <div
    className="search__input"
    role="search"
  >
    <input
      aria-label="Search for cards"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<Search /> shoulder render splashText when prop is set 1`] = `
<div
  className="search"
>
  <div
    className="search__splash-text"
  >
    <div>
      A Point-of-Care Reference Library
    </div>
    <div>
      formerly known as 
      <em>
        Paucis Verbis Cards
      </em>
      <br />
      by the 
      <a
        href="https://aliem.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        ALiEM Team
      </a>
       
      and contributors
    </div>
  </div>
  <div
    className="search__input"
    role="search"
  >
    <input
      aria-label="Search for cards"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </div>
</div>
`;
